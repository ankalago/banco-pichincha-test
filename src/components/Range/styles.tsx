import styled from 'styled-components';

const WrapperRange = styled.div`
  display: flex;
  gap: 0.2em;
  font-size: 14px;
  color: #6b7280;
  width: 100%;
  justify-content: space-between;
`

export {
  WrapperRange
}
