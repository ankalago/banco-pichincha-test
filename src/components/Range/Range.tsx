import React, { useEffect, useState } from 'react';
import { WrapperRange } from './styles';

export type PropsRange = {
  name: string
  label: string
  className?: string
  setValue: Function;
  value?: number;
}

const Range: React.FC<PropsRange> = ({ name, label, className, setValue, value = "" }) => {
  const [valueRange, setValueRange] = useState(0)

  const backgroundRangeStart =(min: string | number, max: string | number, value: string | number) => {
    const minimum = Number(min)
    const maximum = Number(max)
    const val = Number(value) || 0
    return (val - minimum) * 100 / (maximum - minimum);
  }

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    let target = e.target
    const valueRange = Number(target.value)
    target.style.backgroundSize = `${backgroundRangeStart(target.min, target.max, target.value)}% 100%`;
    setValueRange(valueRange)
    setValue(name, valueRange)
  }

  useEffect(() => {
    const target = document.getElementById(name)
    if (target) {
      target.style.backgroundSize = `${backgroundRangeStart(0, 100, value)}% 100%`;
    }
  }, [value])

  return (
    <>
      <label htmlFor={name}>{label}</label>
      <WrapperRange>
        <span>0</span>
        <input id={name} type="range" step={1} min={0} max={100} onChange={handleChange} className={className} value={value || valueRange} data-testid={name} />
        <span>100</span>
      </WrapperRange>
    </>
  )
}

export default Range
