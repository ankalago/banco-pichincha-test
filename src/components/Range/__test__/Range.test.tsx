import { screen, render, fireEvent, waitFor } from '@testing-library/react';
import Range from '../Range';

const renderComponent = () =>
  render(
    <Range name="test" label="Test" setValue={jest.fn()} />,
  );

describe('RangeComponent', () => {

  it('should render the component', () => {
    const { container } = renderComponent();
    const wrapper = screen.getByText(/Test/);
    expect(wrapper).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should user click range', () => {
    const { container } = renderComponent();
    const range = container.querySelector('input[type="range"]')!;
    waitFor(() => {
      fireEvent.change(range, { target: { value: 50 } });
    })
  });
})
