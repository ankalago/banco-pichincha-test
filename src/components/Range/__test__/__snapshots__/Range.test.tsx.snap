// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeComponent should render the component 1`] = `
<div>
  <label
    for="test"
  >
    Test
  </label>
  <div
    class="sc-bcXHqe iNQVck"
  >
    <span>
      0
    </span>
    <input
      data-testid="test"
      id="test"
      max="100"
      min="0"
      step="1"
      type="range"
      value="0"
    />
    <span>
      100
    </span>
  </div>
</div>
`;
